Server/src/RedisManager.cpp

```diff
@@ -0,0 +1,236 @@
+    }
+}
+
+/**
+ * ReadBMPTable, there will be dedicated thread be launched inside and monitor corresponding redis table.
+ *
+ * \param [in] tables             table names to be subscribed.
+ */
+void RedisManager::SubscriberWorker(const std::string& table) {
+    try {
+        swss::DBConnector cfgDb("CONFIG_DB", 0, true);
+
+        swss::SubscriberStateTable conf_table(&cfgDb, table);
+        swss::Select s;
+        s.addSelectable(&conf_table);
+
+        while (!exit_) {
+            swss::Selectable *sel;
+            int ret;
+
+            ret = s.select(&sel, BMP_CFG_TABLE_SELECT_TIMEOUT);
+            if (ret == swss::Select::ERROR) {
+                SWSS_LOG_NOTICE("Error: %s!", strerror(errno));
+                continue;
+            }
+            if (ret == swss::Select::TIMEOUT) {
+                continue;
+            }
+
+            swss::KeyOpFieldsValuesTuple kco;
+            conf_table.pop(kco);
+
+            if (std::get<0>(kco) == "SET") {
+                if (std::get<1>(kco) == "true") {
+                    EnableTable(table);
+                }
+                else {
+                    DisableTable(table);
+                    DisconnectBMP();
+                }
+            }
+            else if (std::get<0>(kco) == "DEL")
+            {
+                LOG_ERR("Config should not be deleted");
+            }
+        }
+    }
+    catch (const exception &e) {
+        LOG_ERR("Runtime error: %s", e.what());
+    }
+}
+
+  
+
+/**
+ * ReadBMPTable, there will be dedicated thread be launched inside and monitor corresponding redis table.
+ *
+ * \param [in] tables             table names to be subscribed.
+ */
+bool RedisManager::ReadBMPTable(const std::vector<std::string>& tables) {
+    for (const auto& table : tables) {
+        std::shared_ptr<std::thread> threadPtr = std::make_shared<std::thread>(
+            std::bind(&RedisManager::SubscriberWorker, this, table));
+        threadList_.push_back(threadPtr);
+    }
+    return true;
+}
+
+
+/**
+ * Enable specific Table
+ *
+ * \param [in] table    Reference to table name, like BGP_NEIGHBOR_TABLE/BGP_RIB_OUT_TABLE/BGP_RIB_IN_TABLE
+ */
+bool RedisManager::EnableTable(const std::string & table) {
+    enabledTables_.insert(table);
+    return true;
+}
+
+/**
+ * Enable BGP_Neighbor* Table
+ *
+ * \param [in] table    Reference to table name BGP_NEIGHBOR_TABLE/BGP_RIB_OUT_TABLE/BGP_RIB_IN_TABLE
+ */
+bool RedisManager::DisableTable(const std::string & table) {
+    enabledTables_.erase(table);
+    return ResetBMPTable(table);
+}
+
+
+/**
+ * Reset ResetBMPTable, this will flush redis
+ *
+ * \param [in] table    Reference to table name BGP_NEIGHBOR_TABLE/BGP_RIB_OUT_TABLE/BGP_RIB_IN_TABLE
+ */
+bool RedisManager::ResetBMPTable(const std::string & table) {
+
+    LOG_INFO("RedisManager ResetBMPTable %s", table.c_str());
+    swss::Table stateBMPTable(&stateDb_, table);
+    std::vector<std::string> keys;
+    stateBMPTable.getKeys(keys);
```

Comment:
>keys [](http://example.com/codeflow?start=26&length=4)

If you want to delete a super large table, these 2 lines will be slow and blocking any other processes database operations for several seconds.

Reply:
later we may use dedicated instance to replace STATE_DB, or do you know any other more effecient way to clear db?

Reply:
The redis blocking issue will be blocking issue for this PR.
1. if you stick with reuse existing StateDB, you may check [https://redis.io/docs/latest/commands/scan/](https://redis.io/docs/latest/commands/scan/)

2. Or, use another dedicated redis instance.

2 is a better system design.

Reply:
https://github.com/sonic-net/sonic-buildimage/pull/19016, created new instance for BMP_STATE_DB.